File: AoC_Dotty/src/main/scala/day01/day01.scala
import scala.collection.mutable.HashSet

object day01 {

  val expenses = myLib.inputInts("input01.txt")
  
  def twoSum(nums: List[Int], target: Int): List[Int] = {
    val diffs = HashSet.empty[Int]
    nums.foreach(v =>
      if (diffs.contains(v))
        return List(v, target - v)
      else diffs.addOne(target - v))
    throw IllegalArgumentException(s"Not contains pair of num with sum of $target")
  }
  def threeSum(nums: List[Int], target: Int): List[Int] =
  {
    val ia = nums.sorted
    val len = ia.size
    for (i <- 0 until len - 3) {
      var l = i + 1
      var r = len - 1
      while (l < r) {
        val list = List(ia(i), ia(l), ia(r))
        list.sum match {
          case x if x == target => return list
          case x if x < target => l += 1
          case _ => r -= 1
        }
      }
    }
    throw IllegalArgumentException("No solution")
  }
  
  def main(args: Array[String]): Unit = {
    val res1 = twoSum(expenses, 2020).reduce(_*_)
    Console.println(res1) // 388075
    val res2 = threeSum(expenses, 2020).reduce(_*_)
    Console.println(res2) // 293450526
  }
}

File: AoC_Kotlin/src/main/kotlin/day01/day01.kt
package day01

import InputLib.DayInput

object day01 {
    fun twoSum(nums: IntArray, target: Int): Int {
        val diffs = HashSet<Int>()
        for (num in nums) {
            val diff = target - num
            if (diffs.contains(diff)) return diff * num
            diffs.add(num)
        }
        throw IllegalArgumentException("No solution")
    }
    fun threeSum(nums: IntArray, target: Int): Int {
        val ia = nums.sorted()
        val len = ia.size
        for (i in 0 .. len - 3) {
            var l = i + 1
            var r = len - 1
            while (l < r) {
                val sum = ia[i] + ia[l] + ia[r]
                when {
                    sum == target -> return ia[i] * ia[l] * ia[r]
                    sum < target -> l++
                    else -> r--
                }
            }
        }
        throw IllegalArgumentException("No solution")
    }
}

fun main() {
    val expenses = DayInput.inputIntsArray("input01.txt")
    val target = 2020
    val res1 = day01.twoSum(expenses, target)
    println(res1) // 388075
    val res2 = day01.threeSum(expenses, target)
    println(res2) // 293450526
}